bootstore/src/schemes/v0/state.rs

```diff
@@ -0,0 +1,594 @@
+                        }
+                    }
+                } else {
+                    // We don't have enough shares
+                    Output::none()
+                }
+            }
+            RackSecretState::Computed { .. } => {
+                // This is just a late share response. We've already informed
+                // the API caller and satisfied any pending learn requests.
+                Output::none()
+            }
+        }
+    }
+}
+
+/// Initialization state for tracking rack init on the coordinator node (the one
+/// local to RSS)
+#[derive(Debug)]
+pub enum RackInitState {
+    // This only exists on the RSS sled
+    Running {
+        start: Ticks,
+        total_members: usize,
+        acks: BTreeSet<Baseboard>,
+
+        // Any unacked messages will be re-sent when the destination peer
+        // connects.
+        unacked_envelopes: Vec<Envelope>,
+    },
+    Initialized,
+    Timeout,
+}
+
+impl RackInitState {
+    // Return true if we just completed rack init, false otherwise
+    pub fn on_ack(&mut self, from: Baseboard) -> bool {
+        if let RackInitState::Running {
+            total_members,
+            acks,
+            unacked_envelopes,
+            ..
+        } = self
+        {
+            unacked_envelopes.retain(|envelope| envelope.to != from);
+            acks.insert(from);
+            if acks.len() == *total_members {
+                *self = RackInitState::Initialized;
+                return true;
+            }
+        }
+        false
+    }
+    pub fn timer_expired(&self, now: Ticks, timeout: Ticks) -> bool {
+        if let RackInitState::Running { start, .. } = self {
+            now.saturating_sub(*start) > timeout
+        } else {
+            false
+        }
+    }
+
+    pub fn is_initialized(&self) -> bool {
+        matches!(self, RackInitState::Initialized)
+    }
+
+    pub fn unacked_peers(&self) -> BTreeSet<Baseboard> {
+        if let RackInitState::Running { unacked_envelopes, .. } = self {
+            unacked_envelopes
+                .iter()
+                .map(|envelope| envelope.to.clone())
+                .collect()
+        } else {
+            BTreeSet::new()
+        }
+    }
+}
+
+/// State stored on the M.2s.
+#[derive(Debug, Serialize, Deserialize)]
+pub struct PersistentState {
+    // The generation number for ledger writing purposes on both M.2s
+    pub version: u32,
+    pub fsm_state: PersistentFsmState,
+}
+
+// Based on the state machine of how `PersistentState` gets updated, we can
+// deterministically infer a version.
+impl From<PersistentFsmState> for PersistentState {
+    fn from(fsm_state: PersistentFsmState) -> Self {
+        use PersistentFsmState::*;
+        let version = match &fsm_state {
+            Uninitialized => 0,
+            InitialMember { distributed_shares, .. } => {
+                (1 + distributed_shares.len()).try_into().unwrap()
```

Comment:
Does it matter that this could be ambiguous?  (e.g. before we've distributed any shares, `InitialMember` and `Learning` look the same; after we've distributed one share, `InitialMember` and `Learned` look the same)

Reply:
Also... what is `version` actually doing?  It seems like it conveys the same information as the inner `fsm_state`.

Reply:
No longer relevant as of [98ada27](https://github.com/oxidecomputer/omicron/pull/3385/commits/98ada27b53d4429463cde94cc91bedec0bc4ef98)

FYI, though, InitialMember and Learned may look the same but are not. Their `pkg` field contains different types.